### PR TITLE
Add torch.library.register_autograd

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* __->__ #124071
* #124066
* #124065
* #124064
* #123937

Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

### CODE REVIEW

File: torch/library.py [hunk 1]
```diff
@@ -553,6 +554,82 @@ def register(func):
         return register(func)
 
 
+def register_autograd(op: _op_identifier, setup_context_fn: Callable, backward_fn: Callable, /, *, lib=None) -> None:
+    r"""Register a backward formula for this custom op.
+
+    In order for an operator to work with autograd, you need to register
+    a backward formula. There are two pieces to this:
+    1. You must tell us what we need to save from the forward pass for
```
Comment (albanD) [inline torch/library.py:562]: Is setup_context actually a must? I would suggest
1. you must tell us how to compute gradients...
2. If you need any of the value from the forward there, you can use setup_ctx to ... 

WDYT?

File: torch/library.py [hunk 2]
```diff
@@ -553,6 +554,82 @@ def register(func):
...
+    the backward pass. This is the "setup_context" function.
+    2. You must tell us how to compute gradients during the backward pass.
+    This is the "backward" function.
+
+    ``setup_context_fn(ctx, inputs, output)`` runs during the forward pass.
+    Please save quantities needed for backward onto the ``ctx`` object via
```
Comment (albanD) [inline torch/library.py:568]: Not sure if I asked this in another PR but should we explicitly document the ctx here as being a custom Function ctx and link to that doc: https://pytorch.org/docs/stable/autograd.html#context-method-mixins ?

File: torch/library.py (hunk 1 above)
Comment (zou3519) [reply torch/library.py:562]: It's not a must, but most people need it. I'll update the wording to that, but I'm not sure if I should switch the order of the args to register_autograd (it's fine if the user passes None as the setup_context)

File: torch/_library/custom_ops.py [hunk 3]
```diff
@@ -391,7 +391,7 @@ def fake_impl(*args, **kwargs):
 
         lib._register_fake(self._name, fake_impl)
 
-        autograd_impl = _library.autograd.make_autograd_impl(self)
+        autograd_impl = _library.autograd.make_autograd_impl(self._opoverload, self)
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_library/custom_ops.py:394]: Yes, via the power of duck typing

[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_library/custom_ops.py:394]: I'll figure out how to update the typing (we should be able to type this via python protocols)

**Comment by zou3519** [comment]
Improved the docs and updated the typing

File: torch/_library/autograd.py [hunk 4]
```diff
@@ -1,26 +1,39 @@
...
+
+
+@dataclasses.dataclass
+class Info:
+    _setup_context_fn: Optional[Callable]
+    _backward_fn: Optional[Callable]
+
+
+# op actually has type "torch._ops.OpOverload", but we've got some circular import issues.
+def make_autograd_impl(op: Any, info: InfoProtocol) -> Callable:
```
Comment (albanD) [inline torch/_library/autograd.py:22]:
```suggestion
def make_autograd_impl(op: "torch._ops.OpOverload", info: InfoProtocol) -> Callable:
```

That's why strings as types are allowed!

**Review by albanD — APPROVED** [review]
nit on typing

File: torch/_library/autograd.py (hunk 4 above)
Comment (zou3519) [reply torch/_library/autograd.py:22]: mypy is smarter nowadays; it will complain that torch isn't imported if we do "torch._ops.OpOverload" :/

Comment (albanD) [reply torch/_library/autograd.py:22]: Ho and it will complain that torch is not used as well?
We can have an `import torch` at the top of the file right?

Comment (zou3519) [reply torch/_library/autograd.py]: I got away with `from . import _ops`, but `import torch` at the top could trigger circular import issues
